lib/gringotts/gateways/sagepay.ex

```diff
@@ -0,0 +1,316 @@
+  # network request in here, and parse it using another private method called
+  # `respond`.
+
+  # @spec commit(_) :: {:ok | :error, Response}
+  defp commit(:post, endpoint, params, opts) do
+    a_url = @url <> endpoint
+
+    HTTPoison.post(a_url, params, opts)
+    |> respond
+  end
+
+  defp commit_for_key(:post, endpoint, params, opts) do
+    a_url = @url <> endpoint
+
+    response = HTTPoison.post(a_url, params, opts)
+    format_response(response)
+  end
+
+  # Function generate_merchant_key generate a merchant session key that will exist only for 400 seconds
+  # and for 3 wrong card identifiers
+
+  defp generate_merchant_key(opts) do
+    merchant_body = Poison.encode!(%{vendorName: opts[:config].vendor})
+
+    merchant_header = [
+      {"Authorization", "Basic " <> opts[:config].auth_id},
+      {"Content-type", "application/json"}
+    ]
+
+    {:ok, merchant_key} =
+      commit_for_key(:post, "merchant-session-keys", merchant_body, merchant_header)
+
+    merchant_key
+    |> Map.get("merchantSessionKey")
+  end
+
+  # card_params returns credit card details of a customer from Gringotts.Creditcard
+
+  defp card_params(card) do
+    %{
+      "cardDetails" => %{
+        "cardholderName" => CreditCard.full_name(card),
+        "cardNumber" => card.number,
+        "expiryDate" =>
+          (card.month * 100 + card.year)
+          |> Integer.to_string()
+          |> String.pad_leading(4, "0"),
+        "securityCode" => card.verification_code
+      }
+    }
+  end
+
+  # Function generate_card_identifier generate a unique cardidentifier for every transaction
+
+  defp generate_card_identifier(card, merchant_key) do
+    card_header = [
+      {"Authorization", "Bearer " <> merchant_key},
+      {"Content-type", "application/json"}
+    ]
+
+    card = card |> Poison.encode!()
+
+    {:ok, card_identifier} = commit_for_key(:post, "card-identifiers", card, card_header)
+
+    card_identifier
+    |> Map.get("cardIdentifier")
+  end
+
+  # Function transaction_details creates the actual body (details of the customer )of the card 
+  # and with merchant_session_key, card_identifiier ,shipping address of a customer, and
+  # other details and converting the map into keyword list
+
+  defp transaction_details(amount, merchant_key, card_identifiier, opts) do
+    {currency, value} = Money.to_string(amount)
+
+    Poison.encode!(%{
+      "transactionType" => opts[:transactionType],
+      "paymentMethod" => %{
+        "card" => %{
+          "merchantSessionKey" => merchant_key,
+          "cardIdentifier" => card_identifiier,
+          "save" => false
+        }
+      },
+      "vendorTxCode" => opts[:vendorTxCode],
+      "amount" => Kernel.trunc(String.to_float(value)),
+      "currency" => currency,
+      "description" => opts[:description],
+      "customerFirstName" => opts[:customerFirstName],
+      "customerLastName" => opts[:customerLastName],
+      "billingAddress" => %{
```

Comment:
Your docs should say that a `billing_address`, if provided, should be an `Address` struct.